feat: add `math/base/special/factorial2f`

anandkaranubc opened the pull request:
Progresses #649

## Description

> What is the purpose of this pull request?

This pull request:

- Adds the JS and C implementation of `math/base/special/factorial2f`.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #649

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: temp commit

anandkaranubc committed:
feat: temp commit

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
feat: temp commit

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
Merge branch 'develop' into feat/factorial2f

anandkaranubc committed:
feat: add complete implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: passed
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
chore: update copyright years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/factorial2f/README.md:
```diff
@@ -0,0 +1,216 @@
+
+# factorial2f
+
+> Evaluate the [double factorial][double-factorial] function as a single-precision floating-point number.
```
Perhaps this description could be made clearer...?

kgryte replied:
> Evaluate the [double factorial][double-factorial] function for a single-precision floating-point number.

anandkaranubc committed:
docs: fix variable declaration

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/factorial2f/test/test.native.js:
```diff
@@ -0,0 +1,99 @@
+			t.strictEqual( v, expected[ i ], 'returns expected value' );
+		} else {
+			delta = absf( v - expected[i] );
+			tol = 2.4 * EPS * absf( expected[i] );
```
It’s still not entirely clear to me why it’s higher here. I checked, and it’s not due to `CFLAGS`. Seems more like a floating-point arithmetic difference in C v/s JS implementation.

kgryte replied:
Yeah, this doesn't make sense. Something seems off. We are multiplying integers. There shouldn't be precision loss. I am not even sure why we have approx tests in the JS version in the first place.

kgryte replied:
In fact, in the original commit, there weren't approx tests: https://github.com/stdlib-js/stdlib/commit/c0dab036acafd8d1e014ee8c6c8d54d3616fd782

kgryte replied:
Okay. I see. You've added approx tests for the single-precision versions. Not sure why that is necessary.

kgryte replied:
Oh, right. Your fixtures are evaluated in double-precision. You need to rethink your fixture generation strategy.

anandkaranubc replied:
Oops, I misinterpreted the [docs](https://docs.scipy.org/doc/scipy-1.15.2/reference/generated/scipy.special.factorial2.html). Turns out it was referring to `float64` instead of `float32`.

<img width="729" alt="image" src="https://github.com/user-attachments/assets/01338e45-5e26-4b8f-87ab-5401410ccaca" />

anandkaranubc replied:
> In fact, in the original commit, there weren't approx tests: [c0dab03](https://github.com/stdlib-js/stdlib/commit/c0dab036acafd8d1e014ee8c6c8d54d3616fd782)

The only reason I used fixture generation instead of simply copying and truncating the JSON from `factorial2` was because those values were evaluated in double precision.

anandkaranubc replied:
On a side note, I also tried using the same JSON from `factorial2` and casting the expected values to `float32` using `float64toFloat32` before comparing them to the calculated values. With this change, there's no need for approximate checks in the JS tests, but the native tests still use the same tolerance value of `2.4 * EPS`.

So the same scenario as what we have right now :(

kgryte replied:
The reason for the discrepancy between JS and C is that the intermediate accumulation is happening in different precisions. Currently, in JS, the accumulation is happening in `double`. In C, the accumulation is happening in `float`.

kgryte replied:
If you want to see the same results in JS, you need to cast each intermediate accumulation via `float64ToFloat32`.

anandkaranubc replied:
🙇🏻‍♂️

anandkaranubc committed:
fix: convert the JS result to float32

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/src/main.c:
```diff
@@ -0,0 +1,66 @@
+	if ( n < 0 || !stdlib_base_is_integerf( n ) ) {
+		return 0.0f / 0.0f; // NaN
+	}
+	v = n;
```
`v` is not necessary. We do that in JS in order to cast the implicit `double` input argument to `int32` using an asm type annotation.

anandkaranubc committed:
chore: apply suggestions from review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/factorial2f/test/fixtures/python/runner.py:
````diff
@@ -0,0 +1,71 @@
+    python> gen(x, './data.json')
+    ```
+    """
+    y = factorial2(x, exact=False)
````
```suggestion
    y = factorial2(x, exact=False).astype(np.float32)
```

However, this change still doesn't affect the return values of the test fixtures.

anandkaranubc committed:
test: remove the approx check in JS tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/factorial2f/test/test.native.js:
```diff
@@ -0,0 +1,99 @@
+	expected = data.expected;
+
+	for ( i = 0; i < x.length; i++ ) {
+		v = factorial2f( x[ i ] );
```
```suggestion
		v = factorial2f( x[ i ] );
		expected[ i ] = float64toFloat32( expected[ i ] );
```

No change in `tol` with this update either.

anandkaranubc commented:
@kgryte Sorry if my recent comments cause any confusion. Just wanted to summarize everything I found so far.

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/src/main.c:
```diff
@@ -0,0 +1,64 @@
+	}
+	out = 1.0f;
+	for ( i = n; i >= last; i -= 2 ) {
+		out *= i;
```
In C, this is equivalent to `out = out * (float)i;`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/lib/main.js:
```diff
@@ -0,0 +1,93 @@
+	}
+	out = 1;
+	for ( i = v|0; i >= last; i -= 2|0 ) {
+		out *= i|0; // asm type annotation
```
In JS, this is equivalent to `out = out * (double)i;`.

anandkaranubc committed:
fix: convert the accumulated result to float32

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
test: revert the changes to include approx tests in JS

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: passed
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
Merge branch 'develop' into feat/factorial2f

anandkaranubc committed:
fix: correct imports to use `max-nth-double-factorial` constant

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/README.md:
```diff
@@ -0,0 +1,216 @@
+
+# factorial2f
+
+> Evaluate the [double factorial][double-factorial] function for a single-precision floating-point number.
```
```suggestion
> Evaluate the [double factorial][double-factorial] function as a single-precision floating-point number.
```

kgryte committed:
docs: update desc

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/lib/index.js:
```diff
@@ -0,0 +1,52 @@
+'use strict';
+
+/**
+* Evaluate the double factorial function for a single-precision floating-point number.
```
```suggestion
* Evaluate the double factorial function as a single-precision floating-point number.
```

kgryte committed:
docs: update desc

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/package.json:
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "@stdlib/math/base/special/factorial2f",
+  "version": "0.0.0",
+  "description": "Evaluate the double factorial function for a single-precision floating-point number.",
```
```suggestion
  "description": "Evaluate the double factorial function as a single-precision floating-point number.",
```

kgryte committed:
docs: update desc

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/lib/main.js:
```diff
@@ -0,0 +1,94 @@
+	}
+	out = float64ToFloat32( 1 );
+	for ( i = v|0; i >= last; i -= 2|0 ) {
+		out *= i|0; // asm type annotation
+		out = float64ToFloat32( out );
+	}
+	return float64ToFloat32( out );
```
```suggestion
		out = float64ToFloat32( out * i );
	}
	return out;
```

It is fine to omit the asm type annotation here. This is only really useful when we can guarantee that `out` is an SMI.

kgryte committed:
refactor: simplify expressions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/src/main.c:
```diff
@@ -0,0 +1,64 @@
+	int32_t i;
+	float out;
+
+	if ( stdlib_base_is_nanf( n ) ) {
+		return 0.0f / 0.0f; // NaN
+	}
```
```suggestion
```

This branch is not possible. `n` is `int32_t`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/src/main.c:
```diff
@@ -0,0 +1,64 @@
+	if ( n > STDLIB_CONSTANT_FLOAT32_MAX_NTH_DOUBLE_FACTORIAL ) {
+		return STDLIB_CONSTANT_FLOAT32_PINF;
+	}
+	if ( n < 0 || !stdlib_base_is_integerf( n ) ) {
```
```suggestion
	if ( n < 0 ) {
```

`n` is `int32_t`. There is never an instance in which `n` is not an integer. This means that the `manifest.json` should be updated, along with the list of includes above.

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/src/main.c:
```diff
@@ -0,0 +1,64 @@
+	if ( n == 0 || n == 1 ) {
+		return 1.0f;
+	}
+	if ( stdlib_base_is_evenf( n ) ) {
```
Don't use `is_evenf` here. We know that `n` is an integer. Use `n&1 == 0`. Meaning, check the last bit. Manifest and list of includes should be updated accordingly.

kgryte replied:
Actually, use `int32-is-even`: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/assert/int32-is-even

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/test/test.js:
```diff
@@ -0,0 +1,109 @@
+			t.strictEqual( v, expected[ i ], 'returns expected value' );
+		} else {
+			delta = absf( v - expected[i] );
+			tol = 2.4 * EPS * absf( expected[i] );
```
We still need approximate equality? Is that because the fixtures are generated at higher precision?

kgryte commented on lib/node_modules/@stdlib/math/base/special/factorial2f/test/fixtures/python/runner.py:
```diff
@@ -0,0 +1,71 @@
+
+def main():
+    """Generate fixture data."""
+    x = np.arange(1, 57, dtype=np.int32)
```
You may be able to adjust the precision here by using `np.float32`. `np.int32` is likely to cause `factorial2` to be evaluated in double-precision. NumPy is typically good about preserving float precision.

anandkaranubc replied:
Thanks @kgryte for the review, and sorry for the mistakes.

Also, I tried using `np.float32`, which requires the `factorial2` function to be called with the additional parameter `extend='complex'`. As the [documentation](https://docs.scipy.org/doc/scipy-1.15.2/reference/generated/scipy.special.factorial2.html) mentions, this changes the values of the double factorial for even integers by reducing them by a factor of sqrt(2/π). But even if we ignore that, the results for odd integers don't match perfectly either.

After a bit of head-twisting and surfing through the source code, I found that SciPy uses the gamma function to approximate the double factorial, rather than the straightforward iterative method we typically use. They have a wrapper function that generalizes the implementation for all types of factorials.

https://github.com/scipy/scipy/blob/0f1fd4a7268b813fa2b844ca6038e4dfdf90084a/scipy/special/_basic.py#L2992-L2997

anandkaranubc replied:
~~I could just generate the expected values locally using MATLAB.~~

Update: No luck with that :(

kgryte replied:
No worries. Just add a note to the tests explaining why we are using approximate equality.

anandkaranubc committed:
fix: use correct imports and remove redundant checks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
test: add comment on the use of approximate equality

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter reviewed (approved):
All review comments have been addressed and the implementation looks good to me, so this PR should be ready to land.
